File: Data Structures & Algorithms/balanced-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if(!root) return true;

        let left = this.height(root.left);
        let right = this.height(root.right);

        if(Math.abs(left-right) >1) return false;

        return this.isBalanced(root.left) && this.isBalanced(root.right);

    }

    height(root){
        if(!root) return 0;

        return 1+ Math.max(this.height(root.left),this.height(root.right));

    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)

        left = 0
        right = 0
        freq = defaultdict(int)
        longest = 0
        max_freq = 0
        
        while right < n:
            freq[s[right]] += 1

            window_length = right - left + 1
            max_freq = max(max_freq, freq[s[right]])

            if window_length - max_freq > k:
                freq[s[left]] -= 1
                left += 1
            longest = max(longest, right - left  + 1)
            right += 1
        return longest




